filecoin/retrieverevents.go

```diff
@@ -0,0 +1,147 @@
+	// RetrievalQueryProgress events occur during the query process, stages
+	// include: connect and query-ask
+	RetrievalQueryProgress(retrievalId uuid.UUID, timestamp time.Time, requestedCid cid.Cid, storageProviderId peer.ID, stage rep.RetrievalEventCode)
+
+	// RetrievalQueryFailure events occur on the failure of querying a storage
+	// provider. A query will result in either a RetrievalQueryFailure or
+	// a RetrievalQuerySuccess event.
+	RetrievalQueryFailure(retrievalId uuid.UUID, timestamp time.Time, requestedCid cid.Cid, storageProviderId peer.ID, err error)
+
+	// RetrievalQueryFailure events occur on successfully querying a storage
+	// provider. A query will result in either a RetrievalQueryFailure or
+	// a RetrievalQuerySuccess event.
+	RetrievalQuerySuccess(retrievalId uuid.UUID, timestamp time.Time, requestedCid cid.Cid, storageProviderId peer.ID, queryResponse retrievalmarket.QueryResponse)
+
+	// RetrievalStart events are fired at the beginning of retrieval-proper, once
+	// a storage provider is selected for retrieval. Note that upon failure,
+	// another storage provider may be selected so multiple RetrievalStart events
+	// may occur for the same retrieval.
+	RetrievalStart(retrievalId uuid.UUID, timestamp time.Time, requestedCid cid.Cid, rootCid cid.Cid, storageProviderId peer.ID)
+
+	// RetrievalProgress events occur during the process of a retrieval. The
+	// Success and failure progress event types are not reported here, but are
+	// signalled via RetrievalSuccess or RetrievalFailure.
+	RetrievalProgress(retrievalId uuid.UUID, timestamp time.Time, requestedCid cid.Cid, storageProviderId peer.ID, stage rep.RetrievalEventCode)
+
+	// RetrievalSuccess events occur on the success of a retrieval. A retrieval
+	// will result in either a RetrievalQueryFailure or a RetrievalQuerySuccess
+	// event.
+	RetrievalSuccess(retrievalId uuid.UUID, timestamp time.Time, requestedCid cid.Cid, storageProviderId peer.ID, retrievalStats filclient.RetrievalStats)
+
+	// RetrievalFailure events occur on the failure of a retrieval. A retrieval
+	// will result in either a RetrievalQueryFailure or a RetrievalQuerySuccess
+	// event.
+	RetrievalFailure(retrievalId uuid.UUID, timestamp time.Time, requestedCid cid.Cid, storageProviderId peer.ID, err error)
+}
+
+type eventManager struct {
+	lk        sync.RWMutex
+	idx       int
+	listeners map[int]RetrievalEventListener
+}
+
+func newEventManager() *eventManager {
+	return &eventManager{listeners: make(map[int]RetrievalEventListener)}
+}
+
+func (em *eventManager) RegisterListener(listener RetrievalEventListener) func() {
+	em.lk.Lock()
+	defer em.lk.Unlock()
+
+	idx := em.idx
+	em.idx++
+	em.listeners[idx] = listener
+
+	// return unregister function
+	return func() {
+		em.lk.Lock()
+		defer em.lk.Unlock()
+		delete(em.listeners, idx)
+	}
+}
+
+func (em *eventManager) fireEvent(cb func(timestamp time.Time, listener RetrievalEventListener)) {
+	timestamp := time.Now()
+	go func() {
+		em.lk.RLock()
```

Comment:
this feels quite racy, cause messages may publish out of order

recommend a single additional go routine that reads messages over a channel and publishes them. You pay need to seperate the read from channel from the publish step to avoid blocking the channel. This is a good reference: https://github.com/filecoin-project/go-statemachine/blob/master/fsm/fsm.go#L149-L173 or you can also use a sync.Cond which I find is a good concurrency primitive for queues.

Reply:
also, unless Listeners get added/removed frequently, I don't think it's a probably to just call the callback inside RLock cause multiple readers can still read at once.